File: clojure/003/003.clj
(defn prime_factors
  ([x]
    (loop [n x p 2 a '()]
      (if (< n (* p p))
        (cons n a)
        (if (= (mod n p) 0)
          (recur (quot n p) p (cons p a))
          (recur n (+ p 1) a)
        )
      )
    )
  )
)

(println (apply max (prime_factors 600851475143)))

File: ruby/003/003.rb
class Integer
  def prime_factors
    n = self

    p = 2

    factors = []

    loop do
      if n < p * p
        factors << n

        break
      elsif n % p == 0
        n /= p

        factors << p
      else
        p += 1
      end
    end

    factors
  end
end

puts 600851475143.prime_factors.max
